src/spac/data_utils.py

```diff
@@ -560,20 +559,22 @@ def downsample_cells(data, annotation, n_samples=None,
                     )
 
             # Concatenate all samples
-            data = pd.concat(samples)
+            output_data = pd.concat(samples)
         else:
-            data = data.groupby(grouping_col).apply(
+            output_data = input_data.groupby(grouping_col).apply(
                 lambda x: x.head(min(n_samples, len(x)))
             ).reset_index(drop=True)
 
         # Remove the combined column if it exists
-        if len(annotation) > 1:
-            data = data.drop(columns=combined_col_name)
+        if len(annotations) > 1:
+            output_data = output_data.drop(columns=combined_col_name)

```

Comment:
@fangliu117 , I blelieve we can keep the combined col in the output data. It would be useful how the stratification took place.

Reply:
Yes, remove these lines.